jina/executors/evaluators/__init__.py

```diff
@@ -0,0 +1,28 @@
+__copyright__ = "Copyright (c) 2020 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+from typing import Iterable
+from jina.proto import jina_pb2
+
+from .. import BaseExecutor
+
+
+class BaseEvaluator(BaseExecutor):
+    """A :class:`BaseEvaluator` evaluates the content of matches against the expected GroundTruth.
+    """
+
+    def __init__(self, id_tag, *args, **kwargs):
+        """"
+        :param id_tag: the key in the tags of documents to identify uniquely a document
+        """
+        super().__init__(*args, **kwargs)
+        self.id_tag = id_tag
+
+    def evaluate(self, matches: Iterable[jina_pb2.Document],
+                 groundtruth: Iterable[jina_pb2.Document], *args, **kwargs) -> float:
+        """"
+        :param matches: the matched documents from the request as matched by jina indexers and rankers
+        :param groundtruth: the expected documents matches sorted as they are expected
+        :return the evaluation metric value for the request document
+        """
+        raise NotImplementedError
```

Comment:
some thoughts:
1.  I don't know how can we will implement the evaluate, will there be a metric@k attached to the `evaluate` method?
2.  `Iterable ` could be tricky, `set` is an `Iterable` object while not keep the order of the document. `Sequence` is something we expected.
3.  A bit curious about the reason why we don't use [abc](https://docs.python.org/3/library/abc.html)

Reply:
Point 1. As I envision it, there will be multiple evaluator executors that wull receive k as parameter and will compute the metricm so one RecallEvaluator, One PrecisionEvaluator, on NDCGEvaluator ...

Point 2. Yes that makes sense it may not be the correct interface.

Point 3. I also would prefer using abc rather than raising exceptions

Reply:
@JoanFM great! Looks good if we can update the type hints. Point 3 could be discussed later :)

Reply:
What is the highest abstraction that respects ordering?

Reply:
frankly I have never done multiple metaclass inherence before, might be a bad idea to do so. I see [this](https://stackoverflow.com/a/57351066) one, might be good for future reference.